README.md


![](https://raw.githubusercontent.com/stewid/SimInf/master/logo/logo.png)

> A flexible and efficient framework for data-driven stochastic disease spread simulations

[![Build Status](https://travis-ci.org/stewid/SimInf.svg?branch=master)](https://travis-ci.org/stewid/SimInf)
[![Build status](https://ci.appveyor.com/api/projects/status/pe68xiu1anxvet2n?svg=true)](https://ci.appveyor.com/project/stewid/SimInf)
[![CRAN status](https://www.r-pkg.org/badges/version/SimInf)](https://CRAN.R-project.org/package=SimInf)
[![CRAN RStudio mirror downloads](https://cranlogs.r-pkg.org/badges/last-month/SimInf)](https://CRAN.R-project.org/package=SimInf)
[![Coverage Status](https://coveralls.io/repos/stewid/SimInf/badge.svg?branch=master&service=github)](https://coveralls.io/github/stewid/SimInf?branch=master)

# SimInf

The package provides an efficient and very flexible framework to
conduct data-driven epidemiological modeling in realistic large scale
disease spread simulations. The framework integrates infection
dynamics in subpopulations as continuous-time Markov chains using the
Gillespie stochastic simulation algorithm and incorporates available
data such as births, deaths and movements as scheduled events at
predefined time-points. Using C code for the numerical solvers and
'OpenMP' (if available) to divide work over multiple processors
ensures high performance when simulating a sample outcome. One of our
design goals was to make the package extendable and enable usage of
the numerical solvers from other R extension packages in order to
facilitate complex epidemiological research. The package contains
template models and can be extended with user-defined models.

## Installation

You can install the released version of `SimInf` from
[CRAN](https://CRAN.R-project.org/package=SimInf)


```r
install.packages("SimInf")
```

or use the `remotes` package to install the development version from
[GitHub](https://github.com/stewid/SimInf)


```r
library(remotes)
install_github("stewid/SimInf")
```

We refer to section 3.1 in the
[vignette](https://CRAN.R-project.org/package=SimInf/vignettes/SimInf.pdf)
for detailed installation instructions.

# Authors

In alphabetical order: Pavol Bauer [![ORCID
iD](https://orcid.org/sites/default/files/images/orcid_16x16.gif)](https://orcid.org/0000-0003-4328-7171),
Robin Eriksson [![ORCID
iD](https://orcid.org/sites/default/files/images/orcid_16x16.gif)](https://orcid.org/0000-0002-4291-712X),
Stefan Engblom [![ORCID
iD](https://orcid.org/sites/default/files/images/orcid_16x16.gif)](https://orcid.org/0000-0002-3614-1732),
and Stefan Widgren [![ORCID
iD](https://orcid.org/sites/default/files/images/orcid_16x16.gif)](https://orcid.org/0000-0001-5745-2284)
**(Maintainer)**

Any suggestions, bug reports, forks and pull requests are
appreciated. Get in touch.

# Acknowledgments

This work was financially supported by the Swedish Research Council
within the UPMARC Linnaeus centre of Excellence (Pavol Bauer, Robin
Eriksson and Stefan Engblom), the Swedish Research Council Formas
(Stefan Engblom and Stefan Widgren), the Swedish Board of Agriculture
(Stefan Widgren), and by the Swedish strategic research program
eSSENCE (Stefan Widgren).

# License

The `SimInf` package is licensed under the
[GPLv3](https://github.com/stewid/SimInf/blob/master/LICENSE).


NEWS
SimInf 6.3.0.9000
-----------------

CHANGES

* The name of the internal function that generated the error has been
  removed from the error message (use 'traceback' to print the call
  stack of the last uncaught error). Additionally, all error messages
  now ends with a '.' (full stop).

BUG FIXES

* Fix storing solution of the state vectors of the first time point
  until after simulated time has passed the the first time point in
  tspan.

SimInf 6.3.0
------------

IMPROVEMENTS

* The 'prevalence' function now supports an additional condition in
  the formula specification to further control the population to
  include in the calculation, see the documentation.

* It is now possible to write the C code snippet of the 'pts_fun'
  argument to 'mparse' as a text string with line breaks instead of as
  a vector of character lines.

* The summary function of a model generated with the 'mparse'
  functionality now includes the propensity of each transition.

BREAKING CHANGES

* Removed the 'U<-' and 'V<-' methods. Use the new 'punchcard<-'
  method instead, see the documentation.

BUG FIXES

* Fix storing solution of the continuous state vector v up to time t,
  but not including t, when t has passed the next time in
  tspan.

* Fix a link-type optimization type mismatch in the internal C
  function 'SimInf_ldata_sp' (detected by the CRAN checks).

SimInf 6.2.0
------------

IMPROVEMENTS

* The 'mparse' model builder functionality has been improved to be
  able to generate models that contain node specific local data and
  continuous state variables. It is also possible to pass the C code
  of the post time step function to 'mparse'. See documentation.

* The package skeleton method has been updated to handle the the
  improved functionlity of 'mparse'.

* Print summary of the local data when the 'ldata' matrix in the model
  has row names.

* It's now possible to pass 'gdata', 'ldata' and 'v0' as data.frames
  to the 'SimInf_model' function.

* Better error message when an invalid rate or a negative state is
  detected during a simulation.

SimInf 6.1.0
------------

IMPROVEMENTS

* Print better error message when processing scheduleed events to
  facilitate debugging.

* Some fixes to the vignette.

* Updated CITATION file.

BUG FIXES

* Fixed broken mparse example.

SimInf 6.0.0
------------

IMPROVEMENTS

* The 'print' and 'summary' methods have been improved to include a
  summary of the data in the model and the number of individuals in
  each compartment.

* Added the 'trajectory' method to facilitate post-processing of
  simulation data. See documentation.

* The event type, when specifying the scheduled events from a
  data.frame, can now be coded either as a numerical value or as a
  character string, see the documentataion for the 'SimInf_events'
  method.

* Now SimInf also includes the All Events Method (AEM) solver (Bauer
  P., Engblom S. (2015) Sensitivity Estimation and Inverse Problems in
  Spatial Stochastic Models of Chemical Kinetics. In: Abdulle A.,
  Deparis S., Kressner D., Nobile F., Picasso M. (eds) Numerical
  Mathematics and Advanced Applications - ENUMATH 2013. Lecture Notes
  in Computational Science and Engineering, vol 103. Springer,
  Cham. Doi: 10.1007/978-3-319-10705-9_51).  A core feature of the AEM
  solver is that reaction events are carried out in channels which
  access private streams of random numbers, in contrast to the
  Gillespie direct method where one use only one stream for all
  events. We have added a new 'solver' argument to the 'run' method.

* The 'package_skeleton' method have been updated to handle the new
  'solver' argument.

* The event type, when specifying the scheduled events from a
  data.frame, can now be coded either as a numerical value or as a
  character string, see the documentataion for the 'SimInf_events'
  method.

* Added the utility function 'select_matrix' to get/set the E matrix
  of the events slot of a SimInf_model object.

* Added the utility function 'shift_matrix' to get/set the N matrix
  of the events slot of a SimInf_model object.

* Added the utility function 'gdata' to get/set parameters in the
  global data vector that is common to all nodes in a SimInf_model
  object.

* Added the utility function 'ldata' to get the local data vector of
  a node in a SimInf_model object.

BREAKING CHANGES

* We have changed 'mparse' method to make it even easier to specify a
  new model for the Siminf framework, see the documentation.

* Removed the following methods: 'U', 'V', 'recovered', 'susceptible',
  and 'infected'. Use the new 'trajectory' method instead, see the
  documentation.

* Renamed the type option 'bnp' to 'nop' in the 'prevalence' method to
  calculate the proportion of nodes with at least one case.

* Removed the 'seed' argument from the 'run' method. Use 'set.seed'
  instead.

SimInf 5.1.0
------------

BUG FIXES

* Added missing calls to 'PROTECT' to protect newly created R objects
  in C code from the garbage collector.

* Protect against invalid transition index in direct SSA.

SimInf 5.0.1
------------

IMPROVEMENTS

* Improvements in the vignette.

SimInf 5.0.0
------------

NEW FEATURES

* Added a vignette that provide a technical description of the
  framework, how to use a predefined model in SimInf, demonstrate a
  case study, and finally show how to extend the framework with a user
  defined model.

* Added the 'pairs' and 'boxplot' methods to visualise the number of
  individuals in each compartment.

* Added 'spaghetti' argument to the plot method to draw one line for
  each node.

* Added the method 'package_skeleton' to create a package skeleton for
  a model depending on SimInf.

CHANGES

* Renamed 'plot' argument 'which' to 'N'.

IMPROVEMENTS

* The core solver now checks the return value from the transition rate
  functions and raise an error if the return value is non-finite or
  less than zero.

* Added example data ('nodes') with a population of 1600 nodes within

R/run.R
## SimInf, a framework for stochastic disease spread simulations
## Copyright (C) 2015  Pavol Bauer
## Copyright (C) 2015 - 2019  Stefan Engblom
## Copyright (C) 2015 - 2019  Stefan Widgren
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see <https://www.gnu.org/licenses/>.

## Use 'R CMD SHLIB' to compile the C code for the model.
do_compile_model <- function(filename)
{
    ## Include directive for "SimInf.h"
    include <- system.file("include", package = "SimInf")
    Sys.setenv(PKG_CPPFLAGS=sprintf("-I%s", shQuote(include)))

    ## Compile the model C code using the running version of R.
    wd <- setwd(dirname(filename))
    cmd <- paste(shQuote(file.path(R.home(component="bin"), "R")),
                 "CMD SHLIB",
                 shQuote(paste0(basename(filename), ".c")))
    compiled <- system(cmd, intern = TRUE)
    setwd(wd)

    lib <- paste0(filename, .Platform$dynlib.ext)
    if (!file.exists(lib))
        stop(compiled, call. = FALSE)

    lib
}

## Check if model contains C code
contains_C_code <- function(model)
{
    if (nchar(paste0(model@C_code, collapse = "\n")))
        return(TRUE)
    FALSE
}

##' Run the SimInf stochastic simulation algorithm
##'
##' @param model The siminf model to run.
##' @param threads Number of threads. Default is NULL, i.e. to use all
##'     available processors.
##' @param solver Which numerical solver to utilize. Default is 'ssm'.
##' @return \code{\link{SimInf_model}} object with result from simulation.
##' @references \itemize{
##'   \item Bauer P, Engblom S, Widgren S
##'   (2016) "Fast Event-Based Epidemiological Simulations on National Scales"
##'   International Journal of High Performance Computing
##'   Applications, 30(4), 438-453. doi:10.1177/1094342016635723
##'
##'   \item Bauer P., Engblom S. (2015) Sensitivity Estimation and
##'   Inverse Problems in Spatial Stochastic Models of Chemical
##'   Kinetics. In: Abdulle A., Deparis S., Kressner D., Nobile F.,
##'   Picasso M. (eds) Numerical Mathematics and Advanced Applications
##'   - ENUMATH 2013. Lecture Notes in Computational Science and
##'   Engineering, vol 103. Springer, Cham. Doi:
##'   10.1007/978-3-319-10705-9_51
##' }
##' @examples
##' ## Create an 'SIR' model with 10 nodes and initialise
##' ## it to run over 100 days.
##' model <- SIR(u0 = data.frame(S = rep(99, 10),
##'                              I = rep(1, 10),
##'                              R = rep(0, 10)),
##'              tspan = 1:100,
##'              beta = 0.16,
##'              gamma = 0.077)
##'
##' ## Run the model and save the result.
##' result <- run(model, threads = 1)
##'
##' ## Plot the proportion of susceptible, infected and recovered
##' ## individuals.
##' plot(result)
setGeneric("run",
           signature = "model",
           function(model,
                    threads = NULL,
                    solver  = c("ssm", "aem"))
               standardGeneric("run"))

##' @rdname run
##' @include SimInf_model.R
##' @export
##' @importFrom methods validObject
setMethod("run",
          signature(model = "SimInf_model"),
          function(model, threads, solver)
          {
              solver <- match.arg(solver)

              ## Check that SimInf_model contains all data structures
              ## required by the siminf solver and that they make sense
              validObject(model);

              if (contains_C_code(model)) {
                  ## Write the C code to a temporary file
                  filename <- tempfile("SimInf-")
                  on.exit(unlink(paste0(filename,
                                        c(".c", ".o", .Platform$dynlib.ex))))
                  writeLines(model@C_code, con = paste0(filename, ".c"))

                  lib <- do_compile_model(filename)
                  dll <- dyn.load(lib)
                  on.exit(dyn.unload(lib), add = TRUE)

                  ## Create expression to parse
                  expr <- ".Call(dll$SimInf_model_run, model, threads, solver)"
              } else {
                  ## The model name
                  name <- as.character(class(model))

                  ## The model C run function
                  run_fn <- paste0(name, "_run")

                  ## Create expression to parse
                  expr <- ".Call(run_fn, model, threads, solver, PACKAGE = 'SimInf')"
              }

              ## Run the model. Re-throw any error without the call
              ## included in the error message to make it cleaner.
              tryCatch(eval(parse(text = expr)), error = function(e) {
                  stop(e$message, call. = FALSE)
              })
          }
)
